packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -233,48 +284,58 @@ contract AmmPool is IBlockReceiver {
 
         // Lock the necessary amounts so we're sure they are available when doing the actual deposit
         for (uint i = 0; i < tokens.length; i++) {
-            require(availableBalance(msg.sender, tokens[i].addr) >= maxAmountsIn[i], "INSUFFICIENT_BALANCE");
-            locked[msg.sender][tokens[i].addr] = locked[msg.sender][tokens[i].addr].add(maxAmountsIn[i]);
+            address token = tokens[i].addr;
+            require(availableBalance(token, msg.sender) >= maxAmountsIn[i], "INSUFFICIENT_BALANCE");
+            locked[token][msg.sender] = locked[token][msg.sender].add(maxAmountsIn[i]);
         }
 
         // Queue the work
         PoolJoin memory join = PoolJoin({
+            owner: msg.sender,
+            fromLayer2: false,
             poolAmountOut: poolAmountOut,
-            maxAmountsIn: maxAmountsIn
+            maxAmountsIn: maxAmountsIn,
+            storageIDs: new uint32[](0)
         });
-        PoolExit memory exit;
         queue.push(QueueItem({
-            owner: msg.sender,
             timestamp: uint64(block.timestamp),
-            workType: WorkType.JOIN,
-            join: join,
-            exit: exit
+            txType: PoolTransactionType.JOIN,
+            txHash: hashPoolJoin(DOMAIN_SEPARATOR, join)
         }));
 
         emit JoinPoolRequested(msg.sender, poolAmountOut, maxAmountsIn);
     }
 
-    function exitPool(uint poolAmountIn, uint96[] calldata minAmountsOut)
+    /// @dev Joins the pool using on-chain funds.
+    /// @param poolAmountIn The number of liquidity tokens that will be burned.
+    /// @param minAmountsOut The minimum amounts that need to be withdrawn when burning
+    ///                      the specified amount of liquidity tokens.
+    function exitPool(uint poolAmountIn, uint96[] calldata minAmountsOut, bool toLayer2)
         external
         online
     {
         require(minAmountsOut.length == tokens.length, "INVALID_DATA");
 
+        // Lock liquidity token - Not needed now with non-transferable liquidity token
```

Comment:
If poolAmountIn and poolAmountOut as requirements cannot be met, how we can fail these Join/Exit?  I don't see related logics.

Reply:
`poolAmountOut` can never fail because that amount will be minted for the user. `poolAmountIn` can fail if the balance for the user is insufficient, is checked here:
```
bool valid = availableBalance(address(this), exit.owner) >= exit.poolAmountIn;
```